File: Python/Sorting/RadixSort.py
'''
Approach:
- The `counting_sort` function is used to sort the input array based on a specific digit's position (given by `exp`).
- The `radix_sort` function repeatedly calls `counting_sort` for each digit's position, starting from the least significant digit to the most significant digit (as controlled by the `exp` variable).
- In each iteration of `counting_sort`, it counts the occurrences of digits at the current position, calculates the cumulative count, and builds the sorted output array.

Time Complexity: The time complexity of Radix Sort is O(n * k), where n is the number of elements in the array, and k is the number of digits in the largest number.

Space Complexity: The space complexity is O(n) as we use extra space for the `output` and `count` arrays.

Sample Input and Output:
Sample Input: [170, 45, 75, 90, 802, 24, 2, 66]
Sorted Output: [2, 24, 45, 66, 75, 90, 170, 802]
'''

def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    # Count occurrences of digits at the current position (exp)
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1

    # Calculate the cumulative count of digits
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build the sorted output array
    i = n - 1
    while i >= 0:
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1

    # Copy the sorted elements back to the original array
    for i in range(n):
        arr[i] = output[i]

def radix_sort(arr):
    max_element = max(arr)
    exp = 1

    # Perform counting sort for each digit's position
    while max_element // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

# Sample input
arr = [170, 45, 75, 90, 802, 24, 2, 66]

# Call radix_sort to sort the array
radix_sort(arr)

# Sample output
print("Sample Input:", [170, 45, 75, 90, 802, 24, 2, 66])
print("Sorted Output:", arr)

File: C++/Sorting/RadixSort.cpp
/*
Explanation:
- The `findMax` function is used to find the maximum element in the array to determine the number of digits in the maximum element.
- `countingSort` is a helper function that performs counting sort for a specific place value (e.g., units, tens, hundreds).
- The `radixSort` function uses counting sort to sort the array for all place values, starting from the least significant digit to the most significant digit.

Time Complexity:
- The time complexity of Radix Sort is O(n * k), where k is the number of digits in the maximum number.

Space Complexity:
- The space complexity is O(n) as it uses additional space to store the output array and counting arrays.

Sample Input:
Sample Input: 170 45 75 90 802 24 2 66

Sample Output:
Sample Output: 2 24 45 66 75 90 170 802

This code performs Radix Sort on the input array and prints the sorted array.
*/

#include <iostream>
#include <vector>
using namespace std;

// A utility function to find the maximum element in the array
int findMax(vector<int>& arr) {
    int max = arr[0];
    for (int i = 1; i < arr.size(); i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    return max;
}

// A function to perform counting sort based on the place value
void countingSort(vector<int>& arr, int exp) {
    const int base = 10; // Base 10 for decimal numbers
    int n = arr.size();
    vector<int> output(n);
    vector<int> count(base, 0);

    // Count occurrences of digits at the specified place value
    for (int i = 0; i < n; i++) {
        count[(arr[i] / exp) % base]++;
    }

    // Calculate the cumulative count
    for (int i = 1; i < base; i++) {
        count[i] += count[i - 1];
    }

    // Build the output array
    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % base] - 1] = arr[i];
        count[(arr[i] / exp) % base]--;
    }

    // Copy the sorted elements back to the original array
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

// Radix Sort function
void radixSort(vector<int>& arr) {
    int max = findMax(arr);

    for (int exp = 1; max / exp > 0; exp *= 10) {
        countingSort(arr, exp);
    }
}

int main() {
    vector<int> arr = {170, 45, 75, 90, 802, 24, 2, 66};
    
    cout << "Sample Input: ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;

    radixSort(arr);

    cout << "Sample Output: ";
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
